File: apps/scanner/app/src/main/java/com/airferry/app/scan/FileNameUtil.kt
package com.airferry.app.scan

import java.io.File

/**
 * Filename helpers shared by the receive / detail / bundle activities.
 *
 * Historically each activity had its own copy of a "sanitize" regex that
 * aggressively stripped everything except ASCII alphanumerics and the CJK
 * Unified Ideographs basic block (U+4E00..U+9FFF). That mangled real-world
 * names — spaces, full-width punctuation (（）), CJK extension characters and
 * most symbols were all replaced with `_`, so a name like `报告 2024.docx`
 * became `报告_2024.docx` and Chinese filenames frequently failed to carry
 * through to share targets (WeChat / QQ / mail clients derive the display
 * name from the FileProvider URI, which used the mangled disk name).
 *
 * Android filesystems (and the SAF) accept a wide range of characters; only
 * a handful are genuinely illegal across filesystems. These helpers strip
 * *only* those, preserving spaces and all CJK characters.
 *
 * The recovered filename is attacker-controllable (decoded from a scanned QR),
 * so [sanitize] also defends against path traversal: it keeps only the final
 * path component, strips separators and control characters, and removes leading
 * dots so the result can never be `.`/`..`/a hidden traversal name. Call sites
 * that write into a directory should use [uniqueTarget], which additionally
 * verifies the resolved path stays inside that directory.
 */
object FileNameUtil {

    /**
     * Strip characters that are illegal in filenames across common filesystems
     * (slash, backslash, colon, asterisk, question mark, double-quote, angle
     * brackets, pipe, and C0 control characters), reduce to the final path
     * component, and drop leading dots. Spaces, full-width punctuation and all
     * Unicode letters (including CJK extension planes) are kept intact.
     *
     * Never returns blank, `.`, or `..` (falls back to `received_file`).
     * Truncates to 200 chars to stay under filesystem path-component limits.
     */
    fun sanitize(name: String): String {
        // Reduce to the final path component first so an attacker-controlled name
        // can't smuggle directory separators / traversal through; then strip
        // illegal chars and leading dots.
        val base = name.substringAfterLast('/').substringAfterLast('\\')
        val stripped = base
            .replace(Regex("[/\\\\:*?\"<>|\\p{Cntrl}]"), "_")
            .trim()
        // Truncate to 200 chars, but never split a UTF-16 surrogate pair at
        // the cut: an orphan low surrogate would land on disk / in share
        // intents as U+FFFD (mirrors FileNameUtil.cs on Windows).
        val cleaned = if (stripped.length > 200) {
            var end = 200
            if (Character.isLowSurrogate(stripped[end])) end -= 1
            stripped.substring(0, end)
        } else {
            stripped
        }
        return cleaned
            .trim()
            .trimStart('.')
            .ifBlank { "received_file" }
    }

    /**
     * Sanitize a logical bundle-relative path while preserving its directory
     * hierarchy. Each component is independently reduced with [sanitize], so
     * traversal (`.`/`..`), separators inside a component and control chars
     * can never escape the logical bundle root.
     */
    fun sanitizeRelativePath(path: String): String {
        val parts = path
            .replace('\\', '/')
            .split('/')
            .filter { it.isNotBlank() && it != "." && it != ".." }
            .map(::sanitize)
        return if (parts.isEmpty()) "received_file" else parts.joinToString("/")
    }

    /**
     * Return a non-existing file in `dir` named `name` (after sanitizing),
     * appending `(1)`, `(2)`, … before the extension on collisions so the
     * original name is never silently overwritten:
     *   报告.docx → 报告.docx
     *   报告.docx (exists) → 报告(1).docx
     *   报告(1).docx (exists) → 报告(2).docx
     *
     * Defensively verifies the resolved path stays inside `dir` (belt-and-
     * suspenders on top of [sanitize]); falls back to a safe name otherwise.
     */
    fun uniqueTarget(dir: File, name: String): File {
        val safe = sanitize(name)
        fun within(f: File): Boolean = try {
            f.canonicalPath.startsWith(dir.canonicalPath + File.separator)
        } catch (_: Exception) {
            false
        }
        val first = File(dir, safe)
        if (!within(first)) return File(dir, "received_file")
        if (!first.exists()) return first
        val dot = safe.lastIndexOf('.')
        val (base, ext) = if (dot in 1 until safe.length - 1) {
            safe.substring(0, dot) to safe.substring(dot)
        } else {
            safe to ""
        }
        var i = 1
        while (i < 10_000 && File(dir, "$base($i)$ext").exists()) i++
        if (i >= 10_000) throw java.io.IOException("目标目录同名文件过多: $safe")
        return File(dir, "$base($i)$ext")
    }

    /**
     * Fixed path under [dir] for a sanitized [name], without appending `(1)`.
     * Use for **share cache** staging: the same logical file may be shared
     * repeatedly, and `uniqueTarget` would otherwise produce `foo(1).ext` on
     * the second share even though the caller overwrites the content.
     * Callers should `copyTo(..., overwrite = true)` onto the returned file.
     */
    fun shareStagingFile(dir: File, name: String): File {
        val safe = sanitize(name)
        fun within(f: File): Boolean = try {
            f.canonicalPath.startsWith(dir.canonicalPath + File.separator)
        } catch (_: Exception) {
            false
        }
        val target = File(dir, safe)
        return if (within(target)) target else File(dir, "received_file")
    }
}

File: apps/windows/AirFerry.Windows/Bundle/FileNameUtil.cs
using System.IO;
using System.Text;

namespace AirFerry.Windows.Bundle;

/// <summary>
/// Filename helpers shared by the receive / detail / bundle views.
/// </summary>
/// <remarks>
/// <para>
/// Mirrors Android's <c>FileNameUtil.kt</c>: strip only characters that are
/// genuinely illegal across filesystems (slash, backslash, colon, asterisk,
/// question mark, double-quote, angle brackets, pipe, C0 control characters),
/// reduce to the final path component, drop leading dots. Spaces, full-width
/// punctuation and all Unicode letters (including CJK extension planes) are
/// kept intact.
/// </para>
/// <para>
/// <b>Windows extras</b> (beyond the Android version):
/// <list type="bullet">
/// <item>Reserved device names (CON, PRN, AUX, NUL, COM1-9, LPT1-9) get a
/// <c>_</c> inserted before the extension so they're no longer treated as
/// devices.</item>
/// <item>Trailing dots/spaces are stripped (Win32 drops them silently, which
/// would change the displayed name).</item>
/// <item>Truncation respects char surrogate pairs to avoid splitting a
/// 4-byte UTF-16 character.</item>
/// </list>
/// </para>
/// <para>
/// The recovered filename is attacker-controllable (decoded from a scanned
/// QR), so <see cref="Sanitize"/> also defends against path traversal.
/// Call sites that write into a directory should use <see cref="UniqueTarget"/>,
/// which additionally verifies the resolved path stays inside that directory.
/// </para>
/// </remarks>
public static class FileNameUtil
{
    private const int MaxComponentChars = 200;
    private const string FallbackName = "received_file";

    /// <summary>
    /// Strip characters that are illegal in filenames across common
    /// filesystems, reduce to the final path component, drop leading dots,
    /// and neutralize Windows reserved device names.
    /// </summary>
    /// <remarks>
    /// Never returns blank, <c>.</c>, or <c>..</c> (falls back to
    /// <c>received_file</c>). Truncates to 200 chars.
    /// </remarks>
    public static string Sanitize(string name)
    {
        // Reduce to the final path component first so an attacker-controlled
        // name can't smuggle directory separators / traversal through.
        string base_ = name;
        int slash = base_.LastIndexOfAny(['/', '\\']);
        if (slash >= 0)
        {
            base_ = base_[(slash + 1)..];
        }

        // Strip genuinely-illegal chars + C0 control chars, replacing with '_'
        // (matches Android's Regex.replace("[/\\:*?\"<>|\\p{Cntrl}]", "_")).
        var sb = new StringBuilder(base_.Length);
        foreach (char c in base_)
        {
            if (IsIllegalFileNameChar(c))
            {
                sb.Append('_');
                continue;
            }
            sb.Append(c);
        }
        string cleaned = sb.ToString().Trim().TrimEnd('.');

        // Truncate without splitting a UTF-16 surrogate pair.
        if (cleaned.Length > MaxComponentChars)
        {
            int cut = MaxComponentChars;
            if (cut > 0 && char.IsLowSurrogate(cleaned[cut]))
            {
                cut--; // keep the high surrogate company.
            }
            cleaned = cleaned[..cut];
        }
        // Trim again after truncation (trailing spaces/dots can reappear).
        cleaned = cleaned.Trim().TrimEnd('.').TrimStart('.');

        if (string.IsNullOrEmpty(cleaned) || cleaned is "." or "..")
        {
            return FallbackName;
        }

        // Windows reserved device names: CON, PRN, AUX, NUL, COM1..9, LPT1..9.
        // Insert "_" before the extension: Win32 ignores the extension while
        // resolving device names, so appending after it ("CON.txt_") would
        // still address CON. "CON_.txt" is an ordinary file.
        return NeutralizeReservedName(cleaned);
    }

    /// <summary>
    /// Sanitize a logical bundle-relative path while preserving directory
    /// hierarchy. Every component is independently sanitized, so traversal
    /// tokens and illegal filename characters cannot escape the bundle root.
    /// The returned logical separator is always '/'.
    /// </summary>
    public static string SanitizeRelativePath(string path)
    {
        string[] parts = path.Replace('\\', '/')
            .Split('/', StringSplitOptions.RemoveEmptyEntries)
            .Where(p => p is not "." and not "..")
            .Select(Sanitize)
            .ToArray();
        return parts.Length == 0 ? FallbackName : string.Join('/', parts);
    }

    /// <summary>
    /// Resolve a non-existing target beneath <paramref name="rootDir"/> while
    /// preserving a sanitized relative directory hierarchy.
    /// </summary>
    public static string UniqueRelativeTarget(string rootDir, string relativeName)
    {
        string safe = SanitizeRelativePath(relativeName);
        string[] parts = safe.Split('/');
        string dir = rootDir;
        for (int i = 0; i < parts.Length - 1; i++)
        {
            string next = Path.Combine(dir, parts[i]);
            if (!IsWithin(rootDir, next))
            {
                return UniqueTarget(rootDir, FallbackName);
            }
            if (PathOccupied(next))
            {
                if (!Directory.Exists(next))
                {
                    throw new IOException($"接收路径的目录位置已被文件占用: {parts[i]}");
                }
                // A lexical StartsWith check does not constrain a junction or
                // symlink's physical target. Never follow attacker-selected
                // bundle components through a pre-existing reparse directory.
                if (IsLinkOrReparsePoint(next))
                {
                    throw new IOException($"接收路径不允许符号链接或重解析目录: {parts[i]}");
                }
            }
            else
            {
                Directory.CreateDirectory(next);
                // Recheck after creation so a concurrent replacement fails
                // closed before a member file is materialized.
                if (IsLinkOrReparsePoint(next))
                {
                    throw new IOException($"接收路径不允许符号链接或重解析目录: {parts[i]}");
                }
            }
            dir = next;
        }
        return UniqueTarget(dir, parts[^1]);
    }

    /// <summary>
    /// Return a non-existing file in <paramref name="dir"/> named
    /// <paramref name="name"/> (after sanitizing), appending <c>(1)</c>,
    /// <c>(2)</c>, … before the extension on collisions so the original name
    /// is never silently overwritten.
    /// </summary>
    public static string UniqueTarget(string dir, string name)
    {
        string safe = Sanitize(name);
        string first = Path.Combine(dir, safe);
        if (!IsWithin(dir, first))
        {
            return Path.Combine(dir, FallbackName);
        }
        if (!PathOccupied(first))
        {
            return first;
        }

        // Split into base + extension for "(N)" insertion.
        string fileName = Path.GetFileName(safe);
        int dot = fileName.LastIndexOf('.');
        string basePart, extPart;
        if (dot >= 1 && dot < fileName.Length - 1)
        {
            basePart = fileName[..dot];
            extPart = fileName[dot..];
        }
        else
        {
            basePart = fileName;
            extPart = string.Empty;
        }

        for (int i = 1; i < 10_000; i++)
        {
            string candidate = Path.Combine(dir, $"{basePart}({i}){extPart}");
            if (!PathOccupied(candidate))
            {
                return candidate;
            }
        }
        throw new IOException($"目标目录同名文件过多: {safe}");
    }

    private static bool PathOccupied(string path)
    {
        if (File.Exists(path) || Directory.Exists(path))
        {
            return true;
        }
        try
        {
            // File.Exists follows links and therefore reports false for a
            // dangling symlink. Attributes/LinkTarget still see the directory
            // entry; treating it as free could make a create/truncate writer
            // follow the link outside the chosen destination.
            _ = File.GetAttributes(path);
            return true;
        }
        catch (FileNotFoundException)
        {
            return HasLinkTarget(path);
        }
        catch (DirectoryNotFoundException)
        {
            return HasLinkTarget(path);
        }
        catch
        {
            // Inaccessible is not equivalent to absent. Fail closed and let
            // the suffix loop choose a different, inspectable path.
            return true;
        }
    }

    private static bool HasLinkTarget(string path)
    {
        try
        {
            return new FileInfo(path).LinkTarget is not null ||
                   new DirectoryInfo(path).LinkTarget is not null;
        }
        catch (FileNotFoundException)
        {
            return false;
        }
        catch (DirectoryNotFoundException)
        {
            return false;
        }
        catch
        {
            return true;
        }
    }

    private static bool IsLinkOrReparsePoint(string path)
    {
        try
        {
            return (File.GetAttributes(path) & FileAttributes.ReparsePoint) != 0 ||
                   new DirectoryInfo(path).LinkTarget is not null;
        }
        catch
        {
            // The caller has just observed/created this directory. Losing the
            // ability to inspect it is a race, not permission to write into it.
            return true;
        }
    }

    private static bool IsIllegalFileNameChar(char c)
    {
        // The Win32 + cross-filesystem illegal set: / \ : * ? " < > | plus all
        // C0 control chars (0x00..0x1F) and DEL (0x7F).
        if (c < 0x20 || c == 0x7F)
        {
            return true;
        }
        return c is '/' or '\\' or ':' or '*' or '?' or '"' or '<' or '>' or '|';
    }

    private static string NeutralizeReservedName(string name)
    {
        // Strip any extension for the reserved-name check (CON.txt is still
        // treated as the CON device by Win32).
        string stem = name;
        int dot = name.IndexOf('.');
        if (dot >= 0)
        {
            stem = name[..dot];
        }
        if (stem.Length is >= 3 and <= 4 && IsReservedStem(stem))
        {
            int insertAt = dot >= 0 ? dot : name.Length;
            string safe = name.Insert(insertAt, "_");
            if (safe.Length > MaxComponentChars)
            {
                int cut = MaxComponentChars;
                if (char.IsLowSurrogate(safe[cut]))
                {
                    cut--;
                }
                safe = safe[..cut];
            }
            return safe;
        }
        return name;
    }

    private static bool IsReservedStem(string stem)
    {
        int len = stem.Length;
        if (len < 3 || len > 4)
        {
            return false;
        }
        // Fixed 3-char names (CON/PRN/AUX/NUL).
        if (len == 3)
        {
            return stem.Equals("CON", StringComparison.OrdinalIgnoreCase)
                || stem.Equals("PRN", StringComparison.OrdinalIgnoreCase)
                || stem.Equals("AUX", StringComparison.OrdinalIgnoreCase)
                || stem.Equals("NUL", StringComparison.OrdinalIgnoreCase);
        }
        // COM1-9 / LPT1-9 (length 4: 3-letter prefix + one digit).
        if (len == 4)
        {
            bool isComOrLpt = stem.StartsWith("COM", StringComparison.OrdinalIgnoreCase)
                || stem.StartsWith("LPT", StringComparison.OrdinalIgnoreCase);
            return isComOrLpt && stem[3] is >= '1' and <= '9';
        }
        return false;
    }

    private static bool IsWithin(string dir, string path)
    {
        try
        {
            string dirCanonical = Path.GetFullPath(dir)
                .TrimEnd(Path.DirectorySeparatorChar) + Path.DirectorySeparatorChar;
            string pathCanonical = Path.GetFullPath(path);
            return pathCanonical.StartsWith(dirCanonical, StringComparison.OrdinalIgnoreCase);
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Soft cap for the in-memory text (copy) UI — mirrors Android
    /// <c>TextLike.MAX_TEXT_UI_BYTES</c> and the web receiver's UTF8_TEXT cap.
    /// Larger text-like files (e.g. multi-MB HTML) stay on the file screen.
    /// </summary>
    public const int MaxTextUiBytes = 256 * 1024;

    /// <summary>
    /// Heuristic: recovered files that should open in the text (copy/share) UI.
    /// Extension-based; mirrors Android <c>TextLike.isTextLikeName</c>.
    /// </summary>
    public static bool IsTextLikeName(string name)
    {
        string baseName = name;
        int slash = baseName.LastIndexOfAny(['/', '\\']);
        if (slash >= 0)
        {
            baseName = baseName[(slash + 1)..];
        }
        int dot = baseName.LastIndexOf('.');
        if (dot <= 0 || dot >= baseName.Length - 1)
        {
            return false;
        }
        string ext = baseName[(dot + 1)..].ToLowerInvariant();
        return TextLikeExtensions.Contains(ext);
    }

    public static bool FitsTextUi(long size) => size >= 0 && size <= MaxTextUiBytes;

    /// <summary>
    /// Decode <paramref name="bytes"/> as UTF-8 only if the sequence is valid and
    /// within <see cref="MaxTextUiBytes"/>. Mirrors Android
    /// <c>TextLike.decodeUtf8Strict</c>.
    /// </summary>
    public static string? DecodeUtf8Strict(byte[] bytes)
    {
        if (bytes.Length > MaxTextUiBytes)
        {
            return null;
        }
        try
        {
            var enc = new System.Text.UTF8Encoding(
                encoderShouldEmitUTF8Identifier: false,
                throwOnInvalidBytes: true);
            return enc.GetString(bytes);
        }
        catch
        {
            return null;
        }
    }

    // Keep in sync with apps/scanner/.../scan/TextLike.kt — only plain-note
    // formats; everything else (HTML, JSON, source code, logs) is a regular
    // file even when technically text.
    private static readonly HashSet<string> TextLikeExtensions = new(StringComparer.Ordinal)
    {
        "txt", "md",
    };
}
